### PR TITLE
[Training] Relax Training Suite

### PR DESCRIPTION
**Pull request by SiriusNEO**
This PR contains a relax training suite to achieve minimum training in Relax including:
- Loss functions
- Optimizers
- Trainer API

It also contains a API fix for Relax block builder, adding argument `name_hint: str` for `emit` and `emit_output`.

Co-authored-by: Yixin Dong <ubospica@gmail.com>

### CODE REVIEW

File: python/tvm/relax/block_builder.py [hunk 1]
```diff
@@ -365,7 +365,7 @@ def dataflow(self) -> DataflowScope:
         """
         return DataflowScope(self)
 
-    def emit(self, expr: Expr) -> Var:
+    def emit(self, expr: Expr, name_hint: str = "") -> Var:
```
Comment (tqchen) [inline python/tvm/relax/block_builder.py:368]: to reduce number of conflicts, please send the name_hint parameter change to upstream

Comment (tqchen) [reply python/tvm/relax/block_builder.py:368]: we don't need to block on upstream being bought up, and can resolve during rebase

**Review by Ubospica — APPROVED** [review]
